consensus/istanbul/backend/val_enode_share.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2017 The Celo Authors
+// This file is part of the celo library.
+//
+// The celo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The celo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the celo library. If not, see <http://www.gnu.org/licenses/>.
+
+package backend
+
+import (
+	"encoding/hex"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus/istanbul"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+// ==============================================
+//
+// define the validator enode share message
+
+type sharedValidatorEnode struct {
+	Address  common.Address
+	EnodeURL string
+	View     *istanbul.View
+}
+
+type valEnodeShareMessage struct {
+	ValEnodes []sharedValidatorEnode
+}
+
+func (sve *sharedValidatorEnode) String() string {
+	return fmt.Sprintf("{Address: %s, EnodeURL: %v, View: %v}", sve.Address.Hex(), sve.EnodeURL, sve.View)
+}
+
+func (sm *valEnodeShareMessage) String() string {
+	outputStr := "{ValEnodes:"
+	for _, valEnode := range sm.ValEnodes {
+		outputStr = fmt.Sprintf("%s %s", outputStr, valEnode.String())
+	}
+	return fmt.Sprintf("%s}", outputStr)
+}
+
+// ==============================================
+//
+// define the functions that needs to be provided for rlp Encoder/Decoder.
+
+// EncodeRLP serializes sm into the Ethereum RLP format.
+func (sm *valEnodeShareMessage) EncodeRLP(w io.Writer) error {
+	return rlp.Encode(w, []interface{}{sm.ValEnodes})
+}
+
+// DecodeRLP implements rlp.Decoder, and load the sm fields from a RLP stream.
+func (sm *valEnodeShareMessage) DecodeRLP(s *rlp.Stream) error {
+	var msg struct {
+		ValEnodes []sharedValidatorEnode
+	}
+
+	if err := s.Decode(&msg); err != nil {
+		return err
+	}
+	sm.ValEnodes = msg.ValEnodes
+	return nil
+}
+
+// This function is meant to be run as a goroutine.  It will periodically gossip validator enode share messages
+// to this node's sentries so that sentries know the enodes of validators
+func (sb *Backend) sendValEnodeShareMsgs() {
+	sb.valEnodeShareWg.Add(1)
+	defer sb.valEnodeShareWg.Done()
+
+	ticker := time.NewTicker(time.Minute)
```

Comment:
I'm worried about only sharing the enode table every minute. Can we share on changes to the table instead?

Reply:
I can do this as a follow up PR after merging the proxy.  It will be easier to wrap it into merging with the val_enode_table changes that Mariano did.

Reply:
Sounds good to me!